ibis/common/egraph.py

```diff
@@ -0,0 +1,432 @@
+from __future__ import annotations
+
+import collections
+import itertools
+import math
+from typing import Any
+
+from ibis.common.collections import DisjointSet
+from ibis.common.graph import Node
+from ibis.util import promote_list
+
+
+class Slotted:
+    """A lightweight alternative to `ibis.common.grounds.Concrete`.
+
+    This class is used to create immutable dataclasses with slots and a precomputed
+    hash value for quicker dictionary lookups.
+    """
+
+    __slots__ = ('__precomputed_hash__',)
+
+    def __init__(self, *args):
+        for name, value in itertools.zip_longest(self.__slots__, args):
+            object.__setattr__(self, name, value)
+        object.__setattr__(self, "__precomputed_hash__", hash(args))
+
+    def __eq__(self, other):
+        if self is other:
+            return True
+        if type(self) is not type(other):
+            return NotImplemented
+        for name in self.__slots__:
```

Comment:
Yeah, basically we want a lightweight and performant base class which support equality and hashing (and preferable "seemingly" immutable). The same applies to my other PR where I have this slotted base class as well. It's not aiming to provide additional feature, we can use `Annotable` for that. 

Perhaps it would be better to just spell the whole object out.